src/components/characters/CardContainer.jsx

```diff
@@ -1,137 +1,79 @@
+import useSWRInfinite from 'swr/infinite';
+import { Button, Box, Typography, CircularProgress } from '@mui/material';
 import Cards from './Cards';
-import { Button, Container, Typography, CircularProgress } from '@mui/material';
+import ErrorDisplay from './ErrorDisplay';
+import LoadingDisplay from './LoadingDisplay';
 
-const API_URL = 'https://rickandmortyapi.com/api/character/?page=1';
+import { commonStyles, CHARACTER_API_URL } from '../../constants';
+import { headerStyles } from './typographyStyles';
 
-const CardContainer = () => {
-  const [characters, setCharacters] = useState([]);
-  const [loading, setLoading] = useState(true);
-  const [error, setError] = useState(null);
-  const [totalCount, setTotalCount] = useState(null);
-  const [nextPageUrl, setNextPageUrl] = useState(API_URL);
+const fetcher = (url) => fetch(url).then((res) => res.json());
 
-  const fetchCharacters = useCallback(async () => {
-    if (!nextPageUrl) {
-      return;
-    }
-    setLoading(true);
-    try {
-      const response = await fetch(nextPageUrl);
-      if (!response.ok) {
-        throw new Error(
-          `Network response was not ok! Status: ${response.status}`,
-        );
-      }
-      const {
-        info: { next: newNext, count },
-        results,
-      } = await response.json();
-      // setTimeout(() => {
-      setCharacters((prevCharacters) => [...prevCharacters, ...results]);
-      if (!totalCount) {
-        setTotalCount(count);
-        console.log('Total count ' + count);
-      }
-      setNextPageUrl(newNext);
-      setLoading(false);
-      // }, 1000);
-    } catch (error) {
-      console.error('Fetch error: ', error);
-      setError(error);
-      setLoading(false);
-    }
-  }, [nextPageUrl]);
+const CardContainer = () => {
+  const getKey = (pageIndex, prevCharacters) => {
+    if (prevCharacters && !prevCharacters.info.next) return null;
+    return prevCharacters ? prevCharacters.info.next : CHARACTER_API_URL;
+  };
 
-  useEffect(() => {
-    fetchCharacters();
-  }, []);
+  const { data, error, size, setSize, isValidating } = useSWRInfinite(
+    getKey,
+    fetcher,
+  );
 
   const handleLoadMore = () => {
-    fetchCharacters();
+    setSize(size + 1);
   };
 
-  console.log('characters ' + characters.length);
-  console.log('loading ' + loading);
-  console.log('error ' + error);
-  console.log('new next ' + nextPageUrl);
+  if (error) return <ErrorDisplay message={error.message} />;
+  if (!data) return <LoadingDisplay />;
 
-  if (error)
-    return (
-      <Container
-        maxWidth='xl'
-        sx={{
-          display: 'flex',
-          flexDirection: 'column',
-          justifyContent: 'center',
-          alignItems: 'center',
-          height: '90vh',
-          marginTop: '20px',
-        }}
-      >
-        <Typography variant='h5'>Error: {error.message}</Typography>
-      </Container>
-    );
+  const characters = data.flatMap((charactersData) => charactersData.results);
```

Comment:
[req] By encapsulating the `characters` state within a `useState` hook and leveraging `useEffect` to update the state when the data changes, you create a clear and efficient state management approach for your component. Hooks must be placed before condition statements
```suggestion
  const [characters, setCharacters] = useState([]);
 
  useEffect(() => {
    if (data) {
      const allCharacters = data.flatMap((charactersData) => charactersData.results)
      setCharacters(allCharacters);
    }
  }, [data]);
```